### ISSUE TITLE
Implement a code formatter for GDScript in the script editor

### ISSUE THREAD

**Issue by Razoric480** [godotengine/godot-proposals#3630]
### Describe the project you are working on

I've been sponsored by GDQuest to make improvements to the godot engine.

### Describe the problem or limitation you are having in your project

See original proposal: This supersedes #1200.

Maintaining a uniform code style between files is time-consuming, and when working with multiple people, it becomes nearly impossible. There is one third party option, https://github.com/Scony/godot-gdscript-toolkit by Scony, but it has to create a syntax tree before every parse - which takes time at boot-up, cannot be run from within Godot, and runs on Python and has to manually text-parse scripts.

If Godot offered the ability to format code, it could be done in-editor and fast (using the built-in parser with C++), and could be out-sourced to the Language Server with the [DocumentFormatting request](https://microsoft.github.io/language-server-protocol/specification#textDocument_formatting) for outside editors.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The new GDScript parser for GDScript 2 does (almost) everything we need it to in order to make a formatter, except for comments and blank lines. Fulfilling this proposal involves:

- A new virtual public function to ScriptLanguage, `format_code(String &p_code)`, in the same vein as `auto_indent_code()` and a new entry in the Edit menu, "Format Code", that calls it. This allows the possibility of maturing Godot's C# editing and any other language bindings to be able to extend it and do their own formatting.

- New Editor Settings related to GDScript formatting:
  - Maximum line width, defaults to 80
  - Choice to add 1 or 2 spaces between functions
  - Format-on-save

- Extend the language server to support DocumentFormatting requests and invoke the formatter through it.

- Add a new option to godot's CLI specifically for batch code formatting. It would take a list of space delineated GDScript files in the project and clean them in series.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

![Screenshot_1](https://user-images.githubusercontent.com/2823241/144629917-f6948318-a9a2-4527-ac34-5d36b7fd1e95.png)

The formatter combines text-parsing to find comments and manual new lines, and the GDScript parser for actual code in order to re-create the script in the [official style guide](https://docs.godotengine.org/en/stable/getting_started/scripting/gdscript/gdscript_styleguide.html). It breaks up lines beyond 80 characters using the:
```gdscript
function(
        param1,
        param2
)
```
syntax and puts arrays/dictionaries/enums on their own lines when they cannot fit, and adds/removes trailing commas in those collections accordingly.

It then re-introduces comments and manual new lines, splits suites up, and otherwise does its utmost best to make it clean and readable while maintaining the coder's organization.

I already have some proof of concept coding working in a branch on my fork, though it's still early goings.

![Screenshot_2](https://user-images.githubusercontent.com/2823241/144632348-2e5afb63-a34d-4441-bc0d-a40875b52267.png)

**Consideration**

It could potentially also re-organize code, grouping constants, signals, exports, vars, onready vars, etc together and putting them in style guide order. This may be more in line with re-factoring tools than pure code formatting, however, so may be best left for those.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Any other option, like Scony's toolkit, requires re-inventing the wheel and comes with other disadvantages, like performance and ease-of-use, not to mention it would be a very large and involved bit of scripting.

### Is there a reason why this should be core and not an add-on in the asset library?

There's a perfectly good parser inside of Godot that we don't have access to outside of C++.

**Comment by Calinou** [comment]
> Add a new option to godot's CLI specifically for batch code formatting. It would take a list of space delineated GDScript files in the project and clean them in series.

Godot has a `--script` CLI option can be combine with `--check-only` to check for syntax. Maybe we can combine that with a `--format` switch?

**Comment by NathanLovato** [comment]
For the cli the `--format` option is great, and yes the values should be space-separated for globbing.

> It could potentially also re-organize code, grouping constants, signals, exports, vars, onready vars, etc together and putting them in style guide order. This may be more in line with re-factoring tools than pure code formatting, however, so may be best left for those.

Such a refactoring tool would be great I think, for many scripts. The official style guide has a particular code order and that ordering can be automated reliably. However, it should definitely at least be a separate tool or menu option. 

But I'm all for it: combined with @Calinou's suggestion to use exit codes, it'd allow to automatically check the code in official demos follows the standard style and code order very efficiently.

**Comment by Amatrelan** [comment]
> The official style guide has a particular code order and that ordering can be automated reliably. However, it should definitely at least be a separate tool or menu option.

I agree refactoring tools would be nice but I think if that kinda tools there should be configuration for things, this is subject what can be too intrusive if done wrong.

Also one more settings I would love to see in this feature would be 
- `Tabs or Spaces indentation`
-  `Indent Width` if spaces is selected.

The reason I don't use those https://github.com/Scony/godot-gdscript-toolkit in formatting is that I cannot choose to use spaces and it overwrites those.

**Comment by Calinou** [comment]
> Also one more settings I would love to see in this feature would be

Both of those settings are already available in the Editor Settings. The code formatter will make use of those :slightly_smiling_face:

**Comment by Razoric480** [comment]
Yep 
![image](https://user-images.githubusercontent.com/2823241/144652727-393f062b-4c89-4aa1-b541-759686981264.png)

**Comment by Amatrelan** [comment]
Ah OK, don't use internal editor at all myself so never looked from there those settings for editor. Hmm would there possibility have mirror settings like if there would be same settings in `Language Server` section?

Because this would clarify for people who use only external editors with LSP.

**Comment by Razoric480** [comment]
`textDocument/formatting` has an `options` parameter with `tabSize` and `insertSpaces`, so presumably it'd be configured on the extension/plugin that supports it. The language server function can then use those as overrides.

**Comment by akien-mga** [comment]
We discussed this in a proposal review meeting and confirmed that we like the idea, it would be great to have this in the next Godot version.

Current status is that this is partly implemented in https://github.com/godotengine/godot/pull/55835, we need @godotengine/gdscript @godotengine/script-editor folks to give it a try and provide some review feedback so @Razoric480 can finalize the implementation when time permits.

**Comment by NathanLovato** [comment]
Razoric's keeping at it slowly while waiting for review, but yeah we'd really need a review from the GDScript team.

**Comment by chexmo** [comment]
This project https://github.com/Scony/godot-gdscript-toolkit is alive and we also have and asset library entry to add it from inside Godot... but i believe it isn't enough!

I believe we should consider including it as a built-in feature ootb. 

Haters like to talk about C# and GDScript, but GDScript is one of the key features of Godot, and the Godot's text editor kinda works very fine for it. The editor just needs some modern tools like this formatter and maybe a member rename tool (i.e. to rename a variable across all the scripts or to rename a function across the project including signals and stuff) 

(PS: idk if there is a feature request for this renamer)

**Comment by Calinou** [comment]
@chexmo Both the [formatter](https://github.com/godotengine/godot/pull/76211) and [renaming](https://github.com/godotengine/godot-proposals/issues/899) functionality are planned to be implemented in core at some point :slightly_smiling_face:

**Comment by m21-cerutti** [comment]
Could be very usefull indeed, and would simplify the process when contributing to demo
https://github.com/godotengine/godot-demo-projects/issues/1090

**Comment by Pheubel** [comment]
I like the idea of having a proper formatter, but the thing that concerns me is it being described as settings for the _editor_, while this makes sense for format on save. in cases where you are working with two different projects that use a slight variation in formatting settings, this could get bothersome.

In other programming languages & projects it is not uncommon to have a separate file that dictates to the formatter what rules it should follow. I do see merit in having a formatter style be applied editor wide, so that it can be shared over different projects, so perhaps both options could exist, where if the project does not have a file with formatting rules, it could fall back on the editor's formatting rules. perhaps it could also be useful to generate a formatting rules file from the editor setting's rules.

**Comment by 616b2f** [comment]
> I like the idea of having a proper formatter, but the thing that concerns me is it being described as settings for the _editor_, while this makes sense for format on save. in cases where you are working with two different projects that use a slight variation in formatting settings, this could get bothersome.
> 
> In other programming languages & projects it is not uncommon to have a separate file that dictates to the formatter what rules it should follow. I do see merit in having a formatter style be applied editor wide, so that it can be shared over different projects, so perhaps both options could exist, where if the project does not have a file with formatting rules, it could fall back on the editor's formatting rules. perhaps it could also be useful to generate a formatting rules file from the editor setting's rules.

Editorconfig (`.editorconfig` which lies usually in the projects root dir) could be used for this use case, let's say with a domain specific prefix like `gdscript_`.

**Comment by uwu-420** [comment]
> In other programming languages & projects it is not uncommon to have a separate file that dictates to the formatter what rules it should follow.

True, but in Godot's case I don't know if making an exception for formatting rules by creating an additional file would be a good idea. Personally, I'd prefer all things to stay in the `project.godot` file. Similar to how some other languages handle this:
- Python: e.g. `[tool.ruff]` table in `pyproject.toml`
- JS: e.g.`"prettier"` entry in `package.json`
- Rust: e.g. `[tool.rustfmt]` table in `cargo.toml`

Although the tools also offer the possibility to put their configuration into separate files like `ruff.toml`, `.prettierrc` or `rustfmt.toml`. But I don't think this is necessary for Godot (unless we get standalone GDScript without the Godot engine 😉). But it should be easier for third-party tools to consume `project.godot` e.g. a third-party formatter should be able to easily extract the formatting config from `project.godot`. Is there already something like that I'm not aware of?

> Editorconfig (.editorconfig which lies usually in the projects root dir) could be used for this use case, let's say with a domain specific prefix like gdscript_.

Please don't use `.editorconfig`. Imo it's a really antiquated and inflexible config file format and there's a good reason that modern formatters do not use it. Just look at the mess that `.editorconfig` is for C# formatting. The only things that imo go in there are general things like this
```editorconfig
[*]
charset = utf-8
trim_trailing_whitespace = true
end_of_line = lf
indent_style = space
insert_final_newline = true
indent_size = 4
```

**Comment by uwu-420** [comment]
Since it was mentioned back in 2021 https://github.com/godotengine/godot-proposals/issues/3630#issuecomment-985738213 I want to point out again to raise awareness, I think it's very important that the formatting functionality should be implemented on the LSP server level such that external editors can easily make use of it. It would be quite annoying (at least for me) to also jump from VS Code back to the Godot editor to run automatic formatting.

**Comment by kasparsiricenko** [comment]
that would be a quality improvement. It would also be nice to add something like Prettier support. Prettier is a well known JS formatter it is opinionated and fairly aggressive but it significantly improves code readability. https://prettier.io/docs/plugins#community-plugins

**Comment by uwu-420** [comment]
For everyone who's passing by, you might want to have a look at what the fine folks at GDQuest made :)
https://github.com/GDQuest/GDScript-formatter

**Comment by TkTech** [comment]
What is the project's stance on project dependencies outside of C++ within the core editor? My gdeye (https://github.com/TkTech/gdeye) seems to address a lot of the issue raised here. It lints, formats, has an LSP and an MCP server, has a `gdeye.toml` so settings are project-specific and shared, isn't opinionated and is highly customizable or usable without a single customization. It's fast, handling ~100k LoC in under 1 second, and can do fun things like automatically add static typing to an entire project.

However, it's rust. It's meant to be used as a library and could be easily integrated with the editor, but it looks to me like the editor has a 100% C++ policy (implicitly or explicitly, couldn't find a policy).

**Comment by Anutrix** [comment]
> What is the project's stance on project dependencies outside of C++ within the core editor? My gdeye (https://github.com/TkTech/gdeye) seems to address a lot of the issue raised here. It lints, formats, has an LSP and an MCP server, has a `gdeye.toml` so settings are project-specific and shared, isn't opinionated and is highly customizable or usable without a single customization. It's fast, handling ~100k LoC in under 1 second, and can do fun things like automatically add static typing to an entire project.
> 
> However, it's rust. It's meant to be used as a library and could be easily integrated with the editor, but it looks to me like the editor has a 100% C++ policy (implicitly or explicitly, couldn't find a policy).

Just my personal opinion, I prefer GDQuest's formatter over this mainly because that is just a formatter and maybe linter, not LSP or something else.
Even if linter, formatter and lsp somehow end in single project due to some overlap, MCP does not. A good set if not a lot many users would strictly not want any MCP logic in their formatter or linter.

Ignoring MCP, some examples of separation for the rest:
Rust has clippy for lint, rustfmt for formatter and rust-analyzer for LSP separate for a good reason.
Python has Ruff for lint, Black for formatter and python-lsp-server/Pylance for LSP.

So I would advice splitting it and keeping it modular if possible. Just my personal opinion.

**Comment by TkTech** [comment]
> > What is the project's stance on project dependencies outside of C++ within the core editor? My gdeye (https://github.com/TkTech/gdeye) seems to address a lot of the issue raised here. It lints, formats, has an LSP and an MCP server, has a `gdeye.toml` so settings are project-specific and shared, isn't opinionated and is highly customizable or usable without a single customization. It's fast, handling ~100k LoC in under 1 second, and can do fun things like automatically add static typing to an entire project.
> > However, it's rust. It's meant to be used as a library and could be easily integrated with the editor, but it looks to me like the editor has a 100% C++ policy (implicitly or explicitly, couldn't find a policy).
> 
> Just my personal opinion, I prefer GDQuest's formatter over this mainly because that is just a formatter and maybe linter, not LSP or something else. Even if linter, formatter and lsp somehow end in single project due to some overlap, MCP does not. A good set if not a lot many users would strictly not want any MCP logic in their formatter or linter.
> 
> Ignoring MCP, some examples of separation for the rest: Rust has clippy for lint, rustfmt for formatter and rust-analyzer for LSP separate for a good reason. Python has Ruff for lint, Black for formatter and python-lsp-server/Pylance for LSP.
> 
> So I would advice splitting it and keeping it modular if possible. Just my personal opinion.

It is modular. It's a library. You can build with or without any of the surface features. https://github.com/TkTech/gdeye?tab=readme-ov-file#building-from-source

> MCP does not

It's a syntax and type aware MCP that returns call graphs, type checking, parses scene files, formats or lints snippets of GDScript for display, etc. There is no "MCP logic" in the formatter or linter. I'm not sure why you would think there would be. I appreciate feedback, but you could at least skim the README first. 

>  Python has Ruff for lint, Black for formatter and python-lsp-server/Pylance for LSP.

Ruff is a linter, formatter, and lsp.

**Comment by AdamGerthel** [comment]
Does [GDScript-formatter](https://github.com/GDQuest/GDScript-formatter) and [gdeye](https://github.com/TkTech/gdeye) use different rules?
